File: 206-reverse-linked-list/reverse-linked-list.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* reverseList(struct ListNode* head) {
    if(head == NULL || head->next == NULL){
        return head;
    }
    struct ListNode *curr = head;
    struct ListNode *prev = NULL;
    struct ListNode *next = NULL;

    while(curr != NULL){
        next = curr->next; //save next
        curr->next = prev; //connect back
        prev = curr; //advance prev to curr
        curr = next; //advance curr
    }
    return prev;
}

File: 206-reverse-linked-list/reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def reverseList(self, head):
        """
        :type head: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        if(head is None or head.next is None):
            return head
        curr = head
        prev = None
        next = None
        
        while(curr is not None):
            next = curr.next    #save next node
            curr.next = prev    #connect to previous node
            prev = curr         #updates the prev node
            curr = next         #updates curr position
        return prev
        

        